### PR TITLE
Configure teleop and autonomous control for elevator

### PR DESCRIPTION
**Pull request by Podcakero**
Elevator should be unable to go too low or high
AND it should go to the different rocket levels autonomously

### CODE REVIEW

**Comment by Podcakero** [comment]
we need to change the CANDigitalInputs to regular DigitalInputs for the upper and lower elevator limit switches

File: src/main/java/frc/robot/RobotMap.java [hunk 1]
```diff
@@ -74,10 +83,28 @@
...
     //Elevator Constants//
-    public static final double ELEVATOR_SPEED_MODIFIER = 1.0;
-    public static final double ELEVATOR_FULL_SPEED = 1.0;
-    public static final double ELEVATOR_ENCODER_TARGET_SPEED_MODIFIER = 100.0; //What the difference between the elevator target and current elevator position will be divided by to determine speed
-    public static final double ELEVATOR_STOP_SPEED = 0.0;
+      //Speed modifers
+      public static final double ELEVATOR_APPROACHING_LOWER_LIMIT_SPEED_MODIFIER = 0.05;
+      public static final double ELEVATOR_FULL_SPEED = 1.0;
+      public static final double ELEVATOR_ENCODER_TARGET_SPEED_MODIFIER = 100.0; //What the difference between the elevator target and current elevator position will be divided by to determine speed
+      public static final double ELEVATOR_STOP_SPEED = 0.0;
```
[the remark this answers is not in the document]
Comment (Podcakero) [reply src/main/java/frc/robot/RobotMap.java]: no...?

File: src/main/java/frc/robot/RobotMap.java [hunk 2]
```diff
@@ -74,10 +83,28 @@
     public static final double DRIVETRAIN_RANGEFINDER_TARGETING_SPEED_MODIFIER = 100.0; //The modifier that the rangefinder value will be divided by to determine speed
 
     //Elevator Constants//
-    public static final double ELEVATOR_SPEED_MODIFIER = 1.0;
-    public static final double ELEVATOR_FULL_SPEED = 1.0;
-    public static final double ELEVATOR_ENCODER_TARGET_SPEED_MODIFIER = 100.0; //What the difference between the elevator target and current elevator position will be divided by to determine speed
-    public static final double ELEVATOR_STOP_SPEED = 0.0;
+      //Speed modifers
+      public static final double ELEVATOR_APPROACHING_LOWER_LIMIT_SPEED_MODIFIER = 0.05;
+      public static final double ELEVATOR_FULL_SPEED = 1.0;
```
[the remark this answers is not in the document]
Comment (Podcakero) [reply src/main/java/frc/robot/RobotMap.java]: no...

File: bin/.gitignore [hunk 3]
```diff
@@ -0,0 +1,121 @@
+
```
[the remark this answers is not in the document]
Comment (Podcakero) [reply bin/.gitignore]: no  the bin folder was its own git repo for some reason

File: src/main/java/frc/robot/RobotMap.java [hunk 4]
```diff
@@ -97,6 +124,9 @@
   //Flags
   public static boolean ballInIntake = true;
 
+  //Non-constant speed modifiers//
+  public static double ELEVATOR_SPEED_MODIFIER = 1.0;
```
[the remark this answers is not in the document]
Comment (Podcakero) [reply src/main/java/frc/robot/RobotMap.java]: depends on how close we are to the bottom of the elevator

File: src/main/java/frc/robot/commands/LowerElevatorCommand.java [hunk 5]
```diff
@@ -30,19 +30,22 @@ protected void initialize()
   protected void execute() 
   {
     RobotMap.elevatorSubsystem.elevatorDown();
+    System.out.println(RobotMap.elevatorEncoder.getPosition());
```
[the remark this answers is not in the document]
Comment (Podcakero) [reply src/main/java/frc/robot/commands/LowerElevatorCommand.java]: this is to get the numbers for the different levels

File: src/main/java/frc/robot/commands/LowerElevatorCommand.java [hunk 6]
```diff
@@ -30,19 +30,22 @@ protected void initialize()
...
-    return RobotMap.elevatorSubsystem.getEncoderPosition() == 0;
+    return RobotMap.elevatorSubsystem.getEncoderPosition() <= RobotMap.ELEVATOR_LOWER_ENCODER_LIMIT;
   }
 
   // Called once after isFinished returns true
   @Override
   protected void end() 
   {
+    RobotMap.elevatorEncoder.setPosition(RobotMap.ELEVATOR_LOWER_ENCODER_LIMIT);
+    RobotMap.elevatorSparkMax.stopMotor();
```
[the remark this answers is not in the document]
Comment (Podcakero) [reply src/main/java/frc/robot/commands/LowerElevatorCommand.java]: set position sets the position of the encoder... as in it tells the encoder "you are now at 0" and the encoder goes "ok"

File: src/main/java/frc/robot/commands/RaiseElevatorCommand.java [hunk 7]
```diff
@@ -28,14 +28,17 @@ public RaiseElevatorCommand(int level, boolean cargo)
...
-      else if (level == 1)
-        encoderTarget = 1.0; //1st level rocket ship encoder value
-      else if (level == 2)
-        encoderTarget = 2.0; //2nd level rocket ship encoder value;
-      else     
-        encoderTarget = 3.0; //3rd level rocket ship encoder value;
+      switch (level) 
+      {
+        case 1:
+          encoderTarget = RobotMap.ELEVATOR_LEVEL_1_CARGO_VALUE;
```
[the remark this answers is not in the document]
Comment (Podcakero) [reply src/main/java/frc/robot/commands/RaiseElevatorCommand.java]: we did... in a different branch

File: src/main/java/frc/robot/subsystems/ElevatorSubsystem.java [hunk 8]
```diff
@@ -54,24 +64,81 @@ public void moveElevator(double speed)
...
   {
-    //Left Trigger goes down
-    if (RobotMap.assistantDriverController.getTriggerAxis(Hand.kLeft) > 0.01)
+    if (RobotMap.elevatorEncoder.getPosition() <= 20.0)
+      RobotMap.ELEVATOR_SPEED_MODIFIER = RobotMap.elevatorEncoder.getPosition() * RobotMap.ELEVATOR_APPROACHING_LOWER_LIMIT_SPEED_MODIFIER;
+    else
+      RobotMap.ELEVATOR_SPEED_MODIFIER = RobotMap.ELEVATOR_FULL_SPEED;
+
+    //Left Trigger goes down ONLY if we are above the lower limit
+    if (RobotMap.assistantDriverController.getTriggerAxis(Hand.kLeft) > RobotMap.ELEVATOR_CONTROLLER_DEADZONE && elevatorCanGoDown())
```
[the remark this answers is not in the document]
Comment (Podcakero) [reply src/main/java/frc/robot/subsystems/ElevatorSubsystem.java]: ok ill change the name

File: src/main/java/frc/robot/RobotMap.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Podcakero) [reply src/main/java/frc/robot/RobotMap.java]: See the elevator subsystem

**Comment by Podcakero** [comment]
Do we want to merge into master?

File: src/main/java/frc/robot/subsystems/ElevatorSubsystem.java [hunk 9]
```diff
@@ -50,9 +48,7 @@ public void elevatorDown()
   {
     moveElevator(-RobotMap.ELEVATOR_AUTONOMOUS_SPEED);
     if (getLowerLimitSwitchState())
-    {
```
Comment (Valance23322) [inline src/main/java/frc/robot/subsystems/ElevatorSubsystem.java]: Why did we remove the brackets here and at line 41/43?

**Pull request closed without merging**